packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1580,34 +1580,112 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     late bool result;
     assert(() {
       if (_debugDisposed) {
-        result = false;
-        return true;
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary('A disposed RenderObject was mutated.'),
+          DiagnosticsProperty<RenderObject>(
+            'The disposed RenderObject was',
+            this,
+            style: DiagnosticsTreeStyle.errorProperty,
+          ),
+        ]);
       }
-      if (owner != null && !owner!.debugDoingLayout) {
+
+      final PipelineOwner? owner = this.owner;
+      // Detached nodes are allowed to mutate and the "can perform mutations`
+      // check will be performed when they re-attach. This assert is
+      // only useful during layout.
+      if (owner == null || !owner.debugDoingLayout) {
         result = true;
         return true;
       }
-      RenderObject node = this;
-      while (true) {
-        if (node._doingThisLayoutWithCallback) {
-          result = true;
-          break;
-        }
-        if (owner != null && owner!._debugAllowMutationsToDirtySubtrees && node._needsLayout) {
-          result = true;
-          break;
-        }
-        if (node._debugMutationsLocked) {
-          result = false;
-          break;
-        }
-        if (node.parent is! RenderObject) {
+
+      RenderObject? activeLayoutRoot = this;
+      while (activeLayoutRoot != null && !activeLayoutRoot._debugMutationsLocked) {
+        final bool mutationsToDirtySubtreesAllowed = activeLayoutRoot.owner?._debugAllowMutationsToDirtySubtrees ?? false;
+        if (mutationsToDirtySubtreesAllowed && activeLayoutRoot._needsLayout) {
           result = true;
-          break;
+          return true;
         }
-        node = node.parent! as RenderObject;
+        final AbstractNode? p = activeLayoutRoot.parent;
+        activeLayoutRoot = p is RenderObject ? p : null;
       }
-      return true;
+
+      // Mutations are not permitted if the PipelineOwner is currently working
+      // on a render subtree that does not include this node. Otherwise,
+      // it's allowed to mutate the subtree within a layout callback.
+      if (activeLayoutRoot != null && activeLayoutRoot._doingThisLayoutWithCallback) {
```

Comment:
Previously, this condition was checked on every node on the path to the activeLayoutRoute, but now it is only checked for the activeLayoutRoot. For my own understanding, why is this change considered ok?

Reply:
The original implementation seems to make more sense. Changed.